0000-access-keys.md

```diff
@@ -0,0 +1,313 @@
+  "methodName": "move",
+  "args": "{...serialized args...}",
+  "amount": 0
+}
+```
+
+In this case the `action` is `call`, so the function checks the `amount` to be within the withdrawal limit, check that the contract name is `chess.app` and if there were the last 3 calls were not in the last 20 blocks issue an async call to the `chess.app`.
+The same `proxy` function in theory can handle other actions, e.g. staking or vesting.
+
+The benefit of having a proxy function on your own account is that it doesn't require additional receipt, because the account's state and the code are available at the transaction verification time.
+
+An example of an access key limited to `proxy` function:
+```rust
+/// vasya.near,3bc2b3b
+AccessKey {
+    nonce: 0,
+    
+    permission: AccessKeyPermission::FunctionCall(FunctionCallPermission {
+        // Allowance can be large enough, since the user is likely trusting the app. 
+        allowance: 1_000_000_000,
+        
+        // This access key restricts access to user's account `vasya.near` contract.
+        // Most likely, the contract code can be deployed and upgraded directly from the wallet.
+        receiver_id: "vasya.near",
+
+        // The method is restricted to proxy, which does all the security checks.
+        method_name: Some("proxy"),
+    }),
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+- Access keys are stored with the `account_id,public_key` key. Where `account_id` and `public_key` are actual Account ID and public keys, and `,` is a separator.
+They should be stored on the same shard as the account.
+- Access keys storage rent should be accounted and paid from the account directly without affecting the allowance.
+- Access keys allowance can exceed the account balance.
+- To validate a transaction signed with the AccessKey, we need to first validate the signature, then fetch the Account and the AccessKey, validate that we have enough funds and verify permissions. 
+- Account creation should now create a full permission access keys with MAX allowance, instead of public keys within the account.
+- SwapKey transaction should just replace the old access key with the given new access key.
+
+### Technical changes
+
+#### `nonce` on the AccessKey level instead of account level
+
+Since access keys can be used by the different people or parties at the same time, we need to be able to 
+have separate nonce for each key instead of a single nonce at the account level.
+With a single nonce on the account level, there is a high probability that 2 apps would use the same nonce for 2 different transactions and one of this transactions would be rejected.
+
+Previously we were ordering transactions by nonce and rejecting transactions with a duplicated or lower nonce.
+With the access key nonce, we still need to order transactions by nonce, but now we need to group them by `account_id,public_key` key instead of just account_id.
```

Comment:
Better group them by hash like suggested above to prevent one access key having constant priority over others.